src/System.Net.Security/src/System/Net/_SecureChannel.cs

```diff
@@ -0,0 +1,1487 @@
+        {
+            get
+            {
+                return _connectionInfo;
+            }
+        }
+
+        internal bool IsValidContext
+        {
+            get
+            {
+                return !(_securityContext == null || _securityContext.IsInvalid);
+            }
+        }
+
+        internal bool IsServer
+        {
+            get
+            {
+                return _serverMode;
+            }
+        }
+
+        internal bool RemoteCertRequired
+        {
+            get
+            {
+                return _remoteCertRequired;
+            }
+        }
+
+        internal void SetRefreshCredentialNeeded()
+        {
+            _refreshCredentialNeeded = true;
+        }
+
+        internal void Close()
+        {
+            if (_securityContext != null)
+            {
+                _securityContext.Dispose();
+            }
+            if (_credentialsHandle != null)
+            {
+                _credentialsHandle.Dispose();
+            }
+        }
+
+        //
+        // SECURITY: we open a private key container on behalf of the caller
+        // and we require the caller to have permission associated with that operation.
+        //
+        private X509Certificate2 EnsurePrivateKey(X509Certificate certificate)
+        {
+            if (certificate == null)
+                return null;
+
+            if (Logging.On) Logging.PrintInfo(Logging.Web, this, SR.Format(SR.net_log_locating_private_key_for_certificate, certificate.ToString(true)));
+
+            try
+            {
+                X509Certificate2 certEx = certificate as X509Certificate2;
+                Type t = certificate.GetType();
+                string certHash = null;
+
+                // Protecting from X509Certificate2 derived classes.
+                if (t != typeof(X509Certificate2))
+                {
+                    if (certificate.Handle != IntPtr.Zero)
+                    {
+                        certEx = new X509Certificate2(certificate.Handle);
+                    }
+                }
+
+                certHash = certEx.Thumbprint;
+
+                if (certEx != null)
+                {
+                    if (certEx.HasPrivateKey)
+                    {
+                        if (Logging.On) Logging.PrintInfo(Logging.Web, this, SR.net_log_cert_is_of_type_2);
+                        return certEx;
+                    }
+
+                    if ((object)certificate != (object)certEx)
+                        certEx.Dispose();
+                }
+
+                X509Certificate2Collection collectionEx;
+
+                // ELSE Try the MY user and machine stores for private key check.
+                // For server side mode MY machine store takes priority.
+                X509Store store = EnsureStoreOpened(_serverMode);
+                if (store != null)
+                {
+                    collectionEx = store.Certificates.Find(X509FindType.FindByThumbprint, certHash, false);
```

Comment:
Stores exist as a concept that we manage; but right now Find always returns no matches.  I can't tell if this code is counting on Find duplicating the certificate handles or not.  (Looping over the collection returns a managed object over the same key handle, calling Find returns a managed object over a different (or at least up-ref'd) key handle).

Since only FindByThumbprint is used I can probably prioritize that to help get things working in a timely manner.